contracts/utils/Checkpoints.sol

```diff
@@ -15,21 +16,233 @@ import "./math/SafeCast.sol";
  * _Available since v4.5._
  */
 library Checkpoints {
-    struct Checkpoint {
-        uint32 _blockNumber;
+    struct Checkpoint224 {
+        uint32 _key;
         uint224 _value;
     }
 
+    function latest(Checkpoint224[] storage self) internal view returns (uint224) {
+        uint256 pos = self.length;
+        return pos == 0 ? 0 : _unsafeAccess(self, pos - 1)._value;
+    }
+
+    function push(
```

Comment:
I don't think it's correct to implement all these methods for `CheckpointN[]` because they rely on invariants that are not guaranteed to be valid for an arbitrary array. The `History` struct encapsulates the array and guarantees the invariants.

I imagine this was done this way to avoid generating a `HistoryN` struct, but I think we need to do that and define these methods for that struct. The name could be something other than "History".

Reply:
We need to come up with a name then.

Reply:
Also, If we have a `HistoryN` structure, then I guess the legacy `History` would include a `History224` instead of a `Checkpoint224[]` array ... which is breaking